zerver/lib/presence.py

```diff
@@ -172,26 +183,59 @@ def get_presence_dict_by_realm(
             "user_profile_id",
             "user_profile__enable_offline_push_notifications",
             "user_profile__date_joined",
+            "last_update_id",
         )
     )
+    # Get max last_update_id from the list.
+    if presence_rows:
+        last_update_id_fetched_by_server: Optional[int] = max(
+            row["last_update_id"] for row in presence_rows
+        )
+    else:
+        # If there are no results, that means that are no new updates to presence
+        # since what the client has last seen. Therefore, returning the same
+        # last_update_id that the client provided is correct.
+        last_update_id_fetched_by_server = last_update_id_fetched_by_client
```

Comment:
We should probably discuss the "echo client" and "possibly returns `None`" details in the API design conversation -- I'm not entirely sure this protocol handles those corner cases in the best possible way.

Reply:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite.20.3E.20response.20last_update_id.20when.20no.20new.20data

Reply:
Looks great!